### PR TITLE
fix(nest): Support RxJS retries and Async Context tracking

### LINKED ISSUE #11923: Interceptors no longer work with RxJS retry operator
### Did you read the migration guide?

- [X] I have read the whole migration guide

### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Potential Commit/PR that introduced the regression

PR #11142

### NestJS version

9.4.2 -> 10.0.3

### Describe the regression

I maintain 2 packages `@geersch/retry` and `@geersch/nestjs-retry`. The former one offers a retry function supporting exponential backoff (powered by RxJS) and the latter allows you to use it via a NestJS interceptor.

```ts
@Controller('cats')
class CatsController {
  @Get()
  @UseInterceptors(new RetryInterceptor(FixedBackoffStrategy))
  findAll(@Headers('x-attempt') header: number): string {
    return 'all the cats';
  }
}
```

This worked fine with NestJS 9.x.x, but with NestJS 10 it breaks and the tests of `@geersch/nestjs-retry` package fail. The retry operator which is eventually piped into the `next.handle()` observable does retry X times, but it appears that the actual controller method is not invoked more than once anymore, breaking the retry mechanism.

### Minimum reproduction code

https://github.com/geersch/retry/pull/5

### Input code

I made PR where I update the dependencies to NestJS 10.0.3. 

https://github.com/geersch/retry/pull/5

The automated tests (Github actions) fail for the `@geersch/nestjs-retry` package. See `retry-interceptor.spec.ts` for the failing tests. It's just a simple controller with a few controller methods and 4 tests which all fail.

You can reproduce this by cloning this repository, switching to the `fix/nestjs10` branch and running the tests.

```
git clone git@github.com:geersch/retry.git
git checkout fix/nestjs10
yarn
```

Be sure to the `build` script (`yarn build`) from the root first before running the tests.

```
yarn workspace @geersch/nestjs-retry test --watch
```

### Expected behavior

I would expect the controller method to be retried when I pipe this into the RxJS retry operator.

### Other

_No response_

### PR DESCRIPTION
**Pull request by nordfjord**
Fixes #11923

### CODE REVIEW

File: packages/core/interceptors/interceptors-consumer.ts [hunk 1]
```diff
@@ -24,16 +24,16 @@ export class InterceptorsConsumer {
     const context = this.createContext(args, instance, callback);
     context.setType<TContext>(type);
 
-    const nextFn = async (i = 0) => {
+    const nextFn = (i: number) => async () => {
       if (i >= interceptors.length) {
-        return this.transformDeferred(next);
+        return defer(() => this.transformDeferred(next));
```
Comment (nordfjord) [inline packages/core/interceptors/interceptors-consumer.ts:29]: By re-introducing the defer here we fix the retries but break the context tracking

Comment (nordfjord) [reply packages/core/interceptors/interceptors-consumer.ts:29]: cc @micalevisk

File: packages/core/interceptors/interceptors-consumer.ts [hunk 2]
```diff
@@ -26,7 +27,7 @@ export class InterceptorsConsumer {
 
     const nextFn = async (i = 0) => {
       if (i >= interceptors.length) {
-        return this.transformDeferred(next);
+        return defer(AsyncResource.bind(() => this.transformDeferred(next)));
```
Comment (nordfjord) [inline packages/core/interceptors/interceptors-consumer.ts:30]: See https://nodejs.org/api/async_context.html#static-method-asyncresourcebindfn-type-thisarg

For information about context tracking and this bind

**Comment by kamilmysliwiec** [comment]
lgtm

**Pull request merged**